### PR TITLE
Swaps performance improvements, logic fixes

### PR DESCRIPTION
**Pull request by christianbaroni**
Fixes APP-1766

## What changed (plus any additional context for devs)
- Improves performance throughout swaps (see commit history for details)
- Main improvements:
  - Optimizes useAnimatedReaction updates — previously there were lots of unnecessary updates due to no dependency array being passed (now passing an empty array in most cases, which limits reactions to changes in the prepare function values)
  - Migrates the token lists from FlashList to FlatList, which seems overall more performant
  - Cleans up the swap confirm button logic to reduce jank

### CODE REVIEW

File: src/reanimated/runOnUIImmediately.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const { runOnUIImmediately } = require('react-native-reanimated/src/threads') as typeof import('react-native-reanimated/src/threads');
```
Comment (walmat) [inline src/reanimated/runOnUIImmediately.ts:2]: I'm assuming there's a reason for this funkiness?

File: src/__swaps__/screens/Swap/components/TokenList/TokenToBuyList.tsx [hunk 2]
```diff
@@ -180,7 +185,15 @@ export const TokenToBuyList = () => {
...
-        renderScrollComponent={ScrollViewWithRef as ComponentType<ScrollViewProps>}
+        renderScrollComponent={props => {
+          return (
+            <Animated.ScrollView
+              // eslint-disable-next-line react/jsx-props-no-spreading
+              {...props}
+              animatedProps={animatedListProps}
+            />
+          );
+        }}
```
Comment (greg-schrammel) [inline src/__swaps__/screens/Swap/components/TokenList/TokenToBuyList.tsx:196]: just make sure this doesn't bring back that bug with favoriting and not rerendering correctly

Comment (christianbaroni) [reply src/__swaps__/screens/Swap/components/TokenList/TokenToBuyList.tsx:196]: Haven't seen any issues, thinking it must have been a FlashList issue

File: src/__swaps__/screens/Swap/hooks/useNativeAssetForChain.ts [hunk 3]
```diff
@@ -21,11 +21,12 @@ export const useNativeAssetForChain = ({ inputAsset }: { inputAsset: SharedValue
 
   useAnimatedReaction(
     () => chainId.value,
-    (currentChainId, previoudChainId) => {
-      if (currentChainId !== previoudChainId) {
+    (currentChainId, previousChainId) => {
+      if (currentChainId !== previousChainId) {
```
Comment (walmat) [inline src/__swaps__/screens/Swap/hooks/useNativeAssetForChain.ts:25]: nice

File: src/__swaps__/screens/Swap/providers/swap-provider.tsx [hunk 4]
```diff
@@ -664,12 +667,18 @@ export const SwapProvider = ({ children }: SwapProviderProps) => {
...
+    () =>
+      SwapWarning.swapWarning.value.type === SwapWarningType.no_quote_available ||
+      SwapWarning.swapWarning.value.type === SwapWarningType.no_route_found ||
+      (internalSelectedInputAsset.value && equalWorklet(internalSelectedInputAsset.value.maxSwappableAmount, '0')),
+    (shouldStop, previous) => {
+      if (shouldStop && previous === false) {
+        SwapInputController.quoteFetchingInterval.stop();
+      }
+    },
+    []
```
Comment (walmat) [inline src/__swaps__/screens/Swap/providers/swap-provider.tsx:681]: does this require cleanup for the areas we were previously calling `.stop()`?

Comment (walmat) [reply src/__swaps__/screens/Swap/providers/swap-provider.tsx:681]: I guess technically calling .stop is a no-op in the case where it's already been stopped so nvm

File: src/__swaps__/screens/Swap/providers/swap-provider.tsx [hunk 5]
```diff
@@ -690,51 +699,55 @@ export const SwapProvider = ({ children }: SwapProviderProps) => {
...
+    ].includes(SwapWarning.swapWarning.value.type);
 
-    if (isQuoteError) {
+    if (quoteUnavailable || isQuoteError) {
       const icon = isReviewSheetOpen ? undefined : '􀕹';
       return { icon, label: isReviewSheetOpen ? quoteError : reviewLabel, disabled: true, type: 'hold' };
     }
 
-    if (!hasEnoughFundsForGas.value) {
+    if (hasEnoughFundsForGas.value === false) {
```
Comment (walmat) [inline src/__swaps__/screens/Swap/providers/swap-provider.tsx:750]: curious if this was because of the `undefined` initialization or another reason

**Review by walmat — APPROVED** [review]
tested android and ios no regressions and seems to be the same or maybe slightly faster on both. Good work! Couple questions mostly for clarity.

File: src/__swaps__/screens/Swap/providers/swap-provider.tsx (hunk 5 above)
Comment (christianbaroni) [reply src/__swaps__/screens/Swap/providers/swap-provider.tsx:750]: Yeah it was, `undefined` represents that we don't yet know if the user has insufficient funds for gas — only `false` should trip this error case

**Pull request merged**